sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/StaticSchemaInference.java

```diff
@@ -88,21 +89,48 @@ enum MethodType {
    */
   public static Schema schemaFromClass(
       Class<?> clazz, FieldValueTypeSupplier fieldValueTypeSupplier) {
+    return schemaFromClass(clazz, fieldValueTypeSupplier, new HashMap<Class, Schema>());
+  }
+
+  private static Schema schemaFromClass(
+      Class<?> clazz,
+      FieldValueTypeSupplier fieldValueTypeSupplier,
+      HashMap<Class, Schema> alreadyVisitedSchemas) {
```

Comment:
Sorry, missed this previously: Inputs should accept the most generic type possible to do what it needs to do. In this case `Map` is preferable, since the input doesn't need to be a `HashMap`. Main concern, since this is an output parameter, is if an `ImmutableMap` is passed, but that'll be caught in tests.

With that said, in line with my comment below about using `synchronized` to make the method thread-safe, you could make the `Map` a static field and use `synchronized` methods to keep it thread-safe. That way the computations are preserved across multiple calls. However, I don't know the general preference in Beam, so it might be good to get another opinion.